src/dos/program_more_output.cpp

```diff
@@ -0,0 +1,513 @@
+{
+	is_definitely_last = false;
+	if (!tabs_remaining) {
+		while (true) {
+			if (!GetCharacterRaw(code, is_definitely_last))
+				return false; // end of data
+
+			if (should_end_on_ctrl_c && code == code_ctrl_c) {
+				if (should_print_ctrl_c) {
+					WriteOut("^C");
+				}
+				return false; // end by CTRL+C
+			}
+
+			// Skip CR/LF characters if requested
+			if (skip_next_cr && code == code_cr) {
+				skip_next_cr = false;
+			} else if (skip_next_lf && code == code_lf) {
+				skip_next_lf = false;
+			} else {
+				break;
+			}
+		}
+
+		// If TAB found, replace it with given number of spaces
+		if (code == '\t') {
+			tabs_remaining         = tab_size;
+			is_tab_definitely_last = is_definitely_last;
+			is_definitely_last     = false;
+		}
+	}
+
+	if (tabs_remaining) {
+		code = ' ';
+		--tabs_remaining;
+		is_definitely_last = is_tab_definitely_last && !tabs_remaining;
+	}
+
+	return true;
+}
+
+uint8_t MoreOutputBase::GetCurrentColumn()
+{
+	const auto page = real_readb(BIOSMEM_SEG, BIOSMEM_CURRENT_PAGE);
+	return CURSOR_POS_COL(page);
+}
+
+uint8_t MoreOutputBase::GetCurrentRow()
+{
+	const auto page = real_readb(BIOSMEM_SEG, BIOSMEM_CURRENT_PAGE);
+	return CURSOR_POS_ROW(page);
+}
+
+// ***************************************************************************
+// Output file/device/stream content via MORE
+// ***************************************************************************
+
+MoreOutputFiles::MoreOutputFiles(Program &program) : MoreOutputBase(program) {}
+
+void MoreOutputFiles::AddFile(const std::string &file_name, const bool is_device)
+{
+	input_files.emplace_back();
+	auto &entry = input_files.back();
+
+	entry.name      = file_name;
+	entry.is_device = is_device;
+}
+
+void MoreOutputFiles::Display()
+{
+	if (SuppressWriteOut("")) {
+		input_files.clear();
+		return;
+	}
+
+	PrepareInternals();
+
+	has_multiple_files  = input_files.size() > 1;
+	should_print_ctrl_c = input_files.empty();
+
+	// Show STDIN or input file(s) content
+	if (input_files.empty())
+		DisplayInputStream();
+	else
+		DisplayInputFiles();
+
+	input_files.clear();
+	WriteOut("\n");
+}
+
+std::string MoreOutputFiles::GetShortName(const std::string &file_name,
+                                          const char *msg_id)
+{
+	assert(msg_id);
+
+	// The shortest name we should be able to display is:
+	// - 3 dots
+	// - 1 path separator
+	// - 8 characters of name
+	// - 1 dot
+	// - 3 characters of extension
+	// This gives 16 characters.
+	// We need to keep the last column free (reduces max length by 1).
+	// Format string contains '%s' (increases max length by 2).
+	constexpr size_t min = 16;
+	const auto max = GetMaxColumns() - std::strlen(MSG_Get(msg_id)) + 1;
+	const auto max_len = std::max(min, max);
+
+	// Nothing to do if file name maches the constraint
+	if (file_name.length() <= max_len)
+		return file_name;
+
+	// We need to shorten the name - try to strip part of the path
+	auto shortened = file_name;
```

Comment:
`auto file_name` would be a lot clearer, but then you need to use `file_path` as the input to the function for things to make sense.